File: text-pattern-finder/net5/StringValidator.cs
using System.Collections.Generic;
using System.Linq;

public class StringValidator
{
    private static Dictionary<char, char> _pairs = new()
    {
        {'(', ')'},
        {'{', '}'},
        {'[', ']'}
    };

    public bool Validate(string text)
    {
        var stack = new Stack<char>();
        foreach (var character in text)
        {
            if(IsOpen(character))
                stack.Push(character);
            else
            {
                var lastCharacter = stack.Pop();
                if (!Closes(lastCharacter, character))
                    return false;
            }
        }

        return stack.Count == 0;
    }

    private bool IsOpen(char character) => new[] { '(', '{', '[' }.Any(x => x == character);
    private bool Closes(char lastCharacter, char character) => character == _pairs[lastCharacter];
}

File: text-pattern-finder/typescript/stringValidator.ts
export class StringValidator {
    private _pairs: Record<string, string> = { '(': ')', '{': '}', '[': ']' };

    validate(text: string): boolean {
        let stack = [];

        for (let character of text) {
            if (this.isOpen(character))
                stack.push(character);
            else {
                var lastCharacter = stack.pop();
                if (!this.closes(lastCharacter!, character))
                    return false;
            }
        }

        return stack.length === 0;
    }

    private isOpen(character: string): boolean {
        return ['(', '{', '['].includes(character);
    }

    private closes(lastCharacter: string, character: string): boolean {
        return character === this._pairs[lastCharacter];
    }
}